### PR TITLE
[Linux] Added warning when /tmp directory is mounted with noexec

### LINKED ISSUE #1693: Could not initialize class com.mojang.blaze3d.systems.RenderSystem starting from 1.19
### Operating System

Linux

### Version of Prism Launcher

Prism Launcher 7.2

### Version of Qt

Qt 5.15.10

### Description of bug

Every version prior to 1.19 works fine, but starting from 1.19 up to 1.20.2 fails to start with the same error message.
I've already reinstalled the launcher, doesn't help.
But everything works fine in the official launcher.

```
Prism Launcher version: 7.2 (Gentoo)


Launched instance in online mode

authserver.mojang.com resolves to:
    [52.85.49.78, 52.85.49.102, 52.85.49.12, 52.85.49.33]

session.minecraft.net resolves to:
    [13.107.246.45, 13.107.213.45, 2620:1ec:46::45, 2620:1ec:bdf::45]

textures.minecraft.net resolves to:
    [13.107.246.45, 13.107.213.45, 2620:1ec:46::45, 2620:1ec:bdf::45]

api.mojang.com resolves to:
    [13.107.213.45, 13.107.246.45, 2620:1ec:bdf::45, 2620:1ec:46::45]


Minecraft folder is:
/usr/home/bit6tream/.local/share/PrismLauncher/instances/1.20.2/.minecraft


Java path is:
/opt/openjdk-bin-17.0.6_p10/bin/java


Java is version 17.0.6, using 64 (amd64) architecture, from Eclipse Adoptium.


AMD Ryzen 5 1600 Six-Core Processor
Advanced Micro Devices, Inc. [AMD/ATI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X] (rev c7)

Subsystem: Micro-Star International Co., Ltd. [MSI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X]

Kernel driver in use: amdgpu

OpenGL version string: 4.6 (Compatibility Profile) Mesa 23.1.8

Main Class:
  net.minecraft.client.main.Main

Native path:
  /usr/home/bit6tream/.local/share/PrismLauncher/instances/1.20.2/natives

Traits:
traits XR:Initial
traits FirstThreadOnMacOS

Libraries:
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-glfw-natives-linux/3.3.2/lwjgl-glfw-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-glfw/3.3.2/lwjgl-glfw-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-jemalloc-natives-linux/3.3.2/lwjgl-jemalloc-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-jemalloc/3.3.2/lwjgl-jemalloc-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-natives-linux/3.3.2/lwjgl-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-openal-natives-linux/3.3.2/lwjgl-openal-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-openal/3.3.2/lwjgl-openal-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-opengl-natives-linux/3.3.2/lwjgl-opengl-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-opengl/3.3.2/lwjgl-opengl-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-stb-natives-linux/3.3.2/lwjgl-stb-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-stb/3.3.2/lwjgl-stb-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-tinyfd-natives-linux/3.3.2/lwjgl-tinyfd-natives-linux-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl-tinyfd/3.3.2/lwjgl-tinyfd-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/lwjgl/lwjgl/3.3.2/lwjgl-3.3.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/github/oshi/oshi-core/6.4.5/oshi-core-6.4.5.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/google/code/gson/gson/2.10.1/gson-2.10.1.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/google/guava/guava/32.1.2-jre/guava-32.1.2-jre.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/ibm/icu/icu4j/73.2/icu4j-73.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/authlib/5.0.47/authlib-5.0.47.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/blocklist/1.0.10/blocklist-1.0.10.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/brigadier/1.1.8/brigadier-1.1.8.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/datafixerupper/6.0.8/datafixerupper-6.0.8.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/logging/1.1.1/logging-1.1.1.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/patchy/2.2.10/patchy-2.2.10.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/text2speech/1.17.9/text2speech-1.17.9.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/commons-codec/commons-codec/1.16.0/commons-codec-1.16.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/commons-io/commons-io/2.13.0/commons-io-2.13.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/commons-logging/commons-logging/1.2/commons-logging-1.2.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-buffer/4.1.97.Final/netty-buffer-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-codec/4.1.97.Final/netty-codec-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-common/4.1.97.Final/netty-common-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-handler/4.1.97.Final/netty-handler-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-resolver/4.1.97.Final/netty-resolver-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-transport-classes-epoll/4.1.97.Final/netty-transport-classes-epoll-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-transport-native-epoll/4.1.97.Final/netty-transport-native-epoll-4.1.97.Final-linux-aarch_64.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-transport-native-epoll/4.1.97.Final/netty-transport-native-epoll-4.1.97.Final-linux-x86_64.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-transport-native-unix-common/4.1.97.Final/netty-transport-native-unix-common-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/io/netty/netty-transport/4.1.97.Final/netty-transport-4.1.97.Final.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/it/unimi/dsi/fastutil/8.5.12/fastutil-8.5.12.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/net/java/dev/jna/jna-platform/5.13.0/jna-platform-5.13.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/net/java/dev/jna/jna/5.13.0/jna-5.13.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/commons/commons-compress/1.22/commons-compress-1.22.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/commons/commons-lang3/3.13.0/commons-lang3-3.13.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/httpcomponents/httpclient/4.5.13/httpclient-4.5.13.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/httpcomponents/httpcore/4.4.16/httpcore-4.4.16.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/logging/log4j/log4j-api/2.19.0/log4j-api-2.19.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/logging/log4j/log4j-core/2.19.0/log4j-core-2.19.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/apache/logging/log4j/log4j-slf4j2-impl/2.19.0/log4j-slf4j2-impl-2.19.0.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/joml/joml/1.10.5/joml-1.10.5.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/org/slf4j/slf4j-api/2.0.7/slf4j-api-2.0.7.jar
  /usr/home/bit6tream/.local/share/PrismLauncher/libraries/com/mojang/minecraft/1.20.2/minecraft-1.20.2-client.jar

Native libraries:

Params:
  --username  --version 1.20.2 --gameDir /usr/home/bit6tream/.local/share/PrismLauncher/instances/1.20.2/.minecraft --assetsDir /usr/home/bit6tream/.local/share/PrismLauncher/assets --assetIndex 8 --uuid  --accessToken  --userType  --versionType release

Window size: 854 x 480

Launcher: standard

Java Arguments:
[-Xms512m, -Xmx4096m, -Duser.language=en]


Minecraft process ID: 23529


[20:35:03] [Datafixer Bootstrap/INFO]: 192 Datafixer optimizations took 172 milliseconds
Exception caught from launcher
java.lang.NoClassDefFoundError: Could not initialize class com.mojang.blaze3d.systems.RenderSystem
	at ab.a(SourceFile:66)
	at eqv.a(SourceFile:2500)
	at eqv.a(SourceFile:2494)
	at net.minecraft.client.main.Main.main(SourceFile:223)
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:88)
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:126)
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:71)
Exiting with ERROR
Process exited with code 2.
Clipboard copy at: 06 Oct 2023 20:36:24 +0300
```

### Steps to reproduce

* Start minecraft 1.19 - 1.20.2

### Suspected cause

No idea, 1.18.2 works just fine with the same setup.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by bit69tream**
Minecraft versions starting from 1.19 would not start at all if /tmp is mounted as `noexec`.

Now, Prism Launcher will check if /tmp has `noexec` and if it does, a message pop-up like this would be displayed:

The C API for querying information about mount points is a bit ugly, but the only other option is to either parse `/etc/mtab` file or output of the `mount` command.

### CODE REVIEW

**Comment by Scrumplex** [comment]
Looks like it's safe to assume that LWJGL is going to use `/tmp` by default, as it uses `java.io.tmpdir`. See https://stackoverflow.com/a/1924576 https://github.com/LWJGL/lwjgl3/blob/ad602750ad87939276d1889b28a757061e90715b/modules/lwjgl/core/src/main/java/org/lwjgl/system/Configuration.java#L81-L106

File: launcher/Application.cpp [hunk 1]
```diff
@@ -988,6 +993,36 @@ Application::Application(int& argc, char** argv) : QApplication(argc, argv)
...
+    // notify user if /tmp is mounted with `noexec` (#1693)
+    {
+        FILE* description = setmntent(MOUNTED, "r");
+        mntent* info = nullptr;
+
+        while ((info = getmntent(description)) != nullptr) {
+            std::string_view directory = info->mnt_dir;
+            std::string_view options = info->mnt_opts;
+
+            if (directory == "/tmp" && options.rfind("noexec") != std::string_view::npos) {
```
Comment (Scrumplex) [inline launcher/Application.cpp]: I wonder if we could use Qt APIs to check if the path is executable in general.

I.e. `QFileInfo("/tmp").isExecutable()`

Comment (bit69tream) [reply launcher/Application.cpp]: I've tried searching through QT Api to see if there would be anything suitable, but no luck.
Even `QStorageInfo` class doesn't provide such information.

Comment (Trial97) [reply launcher/Application.cpp]: I understand that maybe QT doesn't have a straightforward way of doing it, but the question remains: 
can't you just look at the folder if it has the executable permission?


> I.e. `QFileInfo("/tmp").isExecutable()`

Additionally, I did see that the C functions that you use are not 100% the same on all platforms(there are cases where the signature differs).

Comment (bit69tream) [reply launcher/Application.cpp]: And directory permission don't depend on mount options, unfortunately.

Comment (bit69tream) [reply launcher/Application.cpp]: > Additionally, I did see that the C functions that you use are not 100% the same on all platforms(there are cases where the signature differs).

Oh yeah, I didn't even notice. OpenBSD has a completely different interface. I'll try fix it today.

Comment (Scrumplex) [reply launcher/Application.cpp]: In that case using `statvfs()` might be a better choice here. See https://www.man7.org/linux/man-pages/man3/statvfs.3.html

That way we don't have to deal with any string operations and can just use `f_flag & ST_NOEXEC` as the condition

Comment (bit69tream) [reply launcher/Application.cpp]: Thanks, `statvfs` indeed seems like an overall better option. It is also consistent across Linux, FreeBSD and OpenBSD!

**Review by Trial97 — APPROVED** [review]

**Comment by Scrumplex** [comment]
Isn't `statvfs` available on the other two platforms? I thought it was :O

**Comment by bit69tream** [comment]
> Isn't statvfs available on the other two platforms? I thought it was :O

It is, but there are only `ST_RDONLY` and `ST_NOSUID` flags on both OpenBSD and FreeBSD.

**Review by getchoo — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**